File: src/cs/QuoteExtract/QuoteExtract/Extract.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using EnsureThat;
using HtmlAgilityPack;
using System.Text.RegularExpressions;

namespace QuoteExtract
{
    public static class Extract
    {

        public static List<Quote> PullQuotes(string siteUrl)
        {
            Ensure.That(siteUrl).IsNotNullOrWhiteSpace();
                        
            return MapQuotes(new HtmlWeb().Load(siteUrl)?.DocumentNode.SelectSingleNode("//h2"));
        }

        private static List<Quote> MapQuotes(HtmlNode node)
        {
            var jdMeirerQuoteRx = new Regex(@"(“.+”)\s(—|–)\s(.+)");
            var result = new List<Quote>();
            var category = "";
            HtmlNode currentNode = node;

            while (currentNode != null)
            {
                switch (currentNode.Name)
                {
                    case "h2":
                        category = HtmlEntity.DeEntitize(currentNode.InnerText);
                        break;
                    case "p":
                        var quote = MapQuote(currentNode, category, jdMeirerQuoteRx);
                        if (quote != null)
                        {
                            result.Add(quote);
                            currentNode = currentNode.NextSibling;
                        }
                        break;
                    default:
                        break;
                }
                currentNode = currentNode.NextSibling;
            }

            return result;
        }

        private static Quote MapQuote(HtmlNode pTag, string category, Regex quoteRegEx)
        {
            var matchedResults = quoteRegEx.Match(HtmlEntity.DeEntitize(pTag.InnerText) + " " + HtmlEntity.DeEntitize(pTag.NextSibling?.InnerText));
            if (matchedResults.Success)
                return new Quote { Category = category, QuoteText = matchedResults.Groups[1].Value, Author = matchedResults.Groups[3].Value };

            return null;
        }
    }
}

File: src/fs/QuoteExtract/Extract.fs
namespace MeiredQuotes.Extract.FS.QuoteExtract

open FSharp.Data
open System.Text.RegularExpressions

type Quote(category, quoteText, author) =
    member this.Category = category
    member this.QuoteText = quoteText
    member this.Author = author

type Extract() =
    static let jdMeirerQuoteRx = new Regex(@"(“.+”)\s(—|–)\s(.+)") 

    static let rec mapQuotes category (quotes:HtmlNode list) =
        match quotes with 
        | head:HtmlNode :: tail 
          when (head.Name() = "h2") -> mapQuotes (head.InnerText()) tail
        | head:HtmlNode :: next:HtmlNode :: tail 
          when (head.Name() = "p" && next.Name() = "p") ->  
               let quoteMatch = jdMeirerQuoteRx.Match((head.InnerText()) + " " + (next.InnerText()))
               if quoteMatch.Success
               then                        
                    new Quote(category, quoteMatch.Groups.[1].Value, quoteMatch.Groups.[3].Value) :: mapQuotes category tail
               else 
                    mapQuotes category (next :: tail)
        | _ :: tail -> 
               mapQuotes category tail
        | [] -> []
    
    static member PullQuotes(siteUrl) =
        if System.String.IsNullOrWhiteSpace siteUrl
        then raise (System.ArgumentException("siteUrl must have a value"))

        let htmlDocument = HtmlDocument.Load(siteUrl)
        
        htmlDocument.Body().Descendants(["h2";"p";]) |> Seq.toList |> mapQuotes "" 